Fix Firefox Bugs

matthewborden opened the pull request:
Fix Firefox issues:
- Not being able to use arrow keys to navigate past equations.
- Fix backspace (again, firefox wasn't buying into any of the old hacks).

Commit:
Remove text nodes within Mathjax equations.

 We don’t know which node the browser will give us, therefore when
checking if it was part of a Mathjax equation (was an SVG element or
script tag), it’s not a bad idea to check that it’s parent wasn’t
either. For example, the browser may have passed us the text node
within the script. When we’ve found that an element was in an SVG, we
iterate up the DOM tree until we find it’s parent.

Commit:
Delete Mathjax equations, when we delete the space infront of them.

We’re going to outright remove an equation, when you delete the space
infront of the equation, because if you delete that space, every
browser freaks the hell out, when you try to focus the cursor into a
div that’s last element is a span (because there is no text node
there).

Commit:
Set the caret to the node two previous to the current node, in Firefox on navigating left.

Firefox, can’t deal with moving the caret, from the front of the span
to the previous element, so we’re going to avoid that all together and
move straight to the end of the 2 x previous node.

Commit:
Use textContent instead of wholeText.

KJTsanaktsidis commented on source/Editor.js:
```diff
@@ -1993,11 +1993,49 @@ var keyHandlers = {
             if (range.commonAncestorContainer && 
                 range.commonAncestorContainer.parentElement && 
-                checkInSVG.test(range.commonAncestorContainer.nodeName)
+                (checkInSVG.test(range.commonAncestorContainer.nodeName) || checkInSVG.test(range.commonAncestorContainer.parentElement.nodeName))
```
Could this not be fixed just by adding more svg tags to the definition of `checkInSVG`, or by checking what kind of document commonAncestorContainer is in? What is this trying to achieve?

matthewborden replied:
This is trying to catch strange cases where text nodes are passed by the browser as commonAncestorContainer. The most common occurrence of this, is that the text node within a script tag will be passed as the commonAncestorContainer. This intends to confirm that we've covered all the bases, in terms of what the browser could send us.

KJTsanaktsidis replied:
Are we talking about text nodes inside the SVG document?

matthewborden replied:
In that case I was talking about a structure that looked like.
    SPAN.mathjax
        -> SCRIPT
             -> TEXT NODE (The browser passes this one).
 ^^ GH removes the indentation of the tree (edit the comment to view)

I don't think it's too offensive to confirm that the parent of what was passed wasn't an SVG element or a Script tag. Those things should be delete anyway shouldn't they?

KJTsanaktsidis replied:
Is the script outside of the svg element? eew.

KJTsanaktsidis replied:
In that case, isn't what you're really interested in whether or not range.commonAncestorContainer is a child of a `span.mathjax`?

matthewborden replied:
Yep. Mathjax is great at generating disgusting things.

KJTsanaktsidis replied:
It'd be good if you can make the condition of this if statement clearly express that; i.e. do something like the while loop below (terminating at the top of the body or something) to see if the commonAncestorContainer is a child of span.mathjax

matthewborden replied:
'top of the body' by this are you referring to the document body? Just wanted to confirm.

KJTsanaktsidis replied:
The iframe body

matthewborden replied:
Sorry, I just re-read you're earlier comments. Yes, I'm interested if the range.commonAncestorContainer is a child of a Mathjax equation. 

The loop terminates at the body now. You're advocating that the condition be something like:

```
if (mathjaxParent.className === "mathjax''){
     if(mathjaxParent.contains(range.commonAncestorContainer)){
           // remove the mathjaxParent.
     }
     break;
}
```

KJTsanaktsidis replied:
More like

``` javascript
var currentSearchNode = range.commonAncestorContainer;
var mathjaxParent = null;
while (currentSearchNode !== self._doc.body) {
    if (currentSearchNode.classList.contains('mathjax')) {
        mathjaxParent = currentSearchNode;
        break;
    }
    currentSearchNode = currentSearchNode.parent;
}

if (mathjaxParent) {
    event.preventDefault();
    detach(mathjaxParent);
}
```

i.e. it looks like you want to prevent default if the common ancestor is in a span.mathjax.

KJTsanaktsidis replied:
It kinda sucks the DOM api hasn't got a better way of doing this.

KJTsanaktsidis replied:
The other way of expressing it might be

``` javascript
var mathjaxes = self._doc.querySelectorAll('span.mathjax');
var mathjaxParent = _.find(mathjaxes, function(mathjaxEl) {
    return mathjaxEl.contains(range.commonAncestorContainer);
});


if (mathjaxParent) {
    event.preventDefault();
    detach(mathjaxParent);
}
```

Excercise for the reader: do this without underscore :(

KJTsanaktsidis replied:
@slashgrin likes the latter. I think I do too.

KJTsanaktsidis commented on source/Editor.js:
```diff
@@ -2104,26 +2142,33 @@ var keyHandlers = {
         var validNodes = !!(anchorNode && anchorPreviousNode);
         var caretAtStartOfNode = selection.anchorOffset === 0;
-
+        var isMathjaxIOS, isMathjaxIE, isMathjax;
```
Why this change?

KJTsanaktsidis commented on source/Editor.js:
```diff
@@ -2104,26 +2142,33 @@ var keyHandlers = {
             leftRange.setEndBefore(anchorPreviousNode);
-            
+
+            if (isGecko && anchorPreviousNode.previousSibling && anchorPreviousNode.previousSibling.textContent) {
```
Checking the node type of `previousSibling` would convey your intent here more clearly

KJTsanaktsidis commented on source/Editor.js:
```diff
@@ -2104,26 +2142,33 @@ var keyHandlers = {
+
             // set a selection.
             self.setSelection(leftRange);
+            return;
```
This looks pretty redundant

Commit:
Remove style redundancy.

The return on line, 2170 isn’t needed and we’re not using the mathjax
names more than once, so we don’t need to define them in advanced.

Commit:
Confirm that previousSibling is a textNode.

This is more descriptive that checking it has textContent.

Commit:
Stop iterating up the DOM tree at the body element of the document.

We do this, so that the loop will terminate, then confirm that we’re
not going to erase the entire body of the document.

Commit:
Delete Mathjax Equations when a descendent is backspaced onto.

On Backspace, search up the DOM to find out if the element being
backspaced onto is a Mathjax equation. If it is, delete the Mathjax
Equation.
